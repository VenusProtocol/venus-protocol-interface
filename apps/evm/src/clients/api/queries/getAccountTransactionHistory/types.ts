import type { ChainId, VToken } from '@venusprotocol/chains';
import type { Token, TxType } from 'types';
import type { Address } from 'viem';
import type { GetPoolsOutput } from '../useGetPools/types';

export interface ApiAccountHistoricalTransaction {
  id: string;
  txHash: string;
  txIndex: number;
  txTimestamp: Date;
  blockNumber: string;
  txType: TxType;
  accountAddress: Address;
  contractAddress: Address;
  amountVTokenMantissa: string | null;
  amountUnderlyingMantissa: string | null;
  underlyingAddress: string;
  underlyingTokenPriceMantissa: string | null;
  chainId: ChainId;
}

export interface AmountTransaction {
  txType: TxType;
  hash: string;
  blockTimestamp: Date;
  blockNumber: string;
  accountAddress: string;
  contractAddress: string;
  chainId: ChainId;
  poolName: string;
  vTokenSymbol: string;
  amountTokens: BigNumber | undefined;
  amountCents: BigNumber | undefined;
  token: Token | VToken;
}

export interface AccountTransactionHistoryApiResponse {
  count: number;
  results: ApiAccountHistoricalTransaction[];
}

export interface GetAccountTransactionHistoryInput {
  accountAddress: string;
  contractAddress: string;
  chainId: ChainId;
  getPoolsData: GetPoolsOutput | undefined;
  type?: number;
  page?: number;
}

export type GetAccountTransactionHistoryOutput = {
  count: number;
  transactions: AmountTransaction[];
};
