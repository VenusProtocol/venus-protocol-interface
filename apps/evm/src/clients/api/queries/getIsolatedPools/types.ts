import type { PoolLens, Prime, ResilientOracle } from 'libs/contracts';
import type { Provider } from 'libs/wallet';
import type { ChainId, Pool, Token } from 'types';

export interface GetIsolatedPoolsInput {
  chainId: ChainId;
  xvs: Token;
  tokens: Token[];
  provider: Provider;
  poolRegistryContractAddress: string;
  vTreasuryContractAddress: string;
  poolLensContract: PoolLens;
  resilientOracleContract: ResilientOracle;
  blocksPerDay?: number;
  primeContract?: Prime;
  accountAddress?: string;
}

export interface GetIsolatedPoolsOutput {
  pools: Pool[];
}
