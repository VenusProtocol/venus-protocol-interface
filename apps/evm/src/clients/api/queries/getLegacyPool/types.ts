import {
  LegacyPoolComptroller,
  Prime,
  ResilientOracle,
  VaiController,
  VenusLens,
} from 'packages/contracts';
import { Pool, Token } from 'types';

export interface GetLegacyPoolInput {
  blocksPerDay: number;
  name: string;
  description: string;
  xvs: Token;
  vai: Token;
  tokens: Token[];
  legacyPoolComptrollerContract: LegacyPoolComptroller;
  venusLensContract: VenusLens;
  resilientOracleContract: ResilientOracle;
  vaiControllerContract: VaiController;
  primeContract?: Prime;
  accountAddress?: string;
}

export interface GetLegacyPoolOutput {
  pool: Pool;
}
