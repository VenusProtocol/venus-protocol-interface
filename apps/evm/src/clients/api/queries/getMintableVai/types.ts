import BigNumber from 'bignumber.js';

import { Vai, VaiController } from 'libs/contracts';

export interface GetMintableVaiInput {
  accountAddress: string;
  vaiControllerContract: VaiController;
  vaiContract: Vai;
}

export interface GetMintableVaiOutput {
  vaiLiquidityMantissa: BigNumber;
  accountMintableVaiMantissa: BigNumber;
}
