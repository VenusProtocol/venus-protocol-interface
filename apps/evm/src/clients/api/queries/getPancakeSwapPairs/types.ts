import { Pair as PSPair, Token as PSToken } from '@pancakeswap/sdk/dist/index.js';
import BigNumber from 'bignumber.js';

import { type Provider } from 'libs/wallet';
import { PSTokenCombination } from 'types';

export interface GetPancakeSwapPairsInput {
  provider: Provider;
  tokenCombinations: PSTokenCombination[];
}

export interface PairAddress {
  tokenCombination: PSTokenCombination;
  address: string;
}

export interface TokenReserve {
  token: PSToken;
  reserveMantissa: BigNumber;
}

export type GetPancakeSwapPairsOutput = {
  pairs: PSPair[];
};
