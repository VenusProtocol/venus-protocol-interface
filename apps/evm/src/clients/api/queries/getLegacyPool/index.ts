import BigNumber from 'bignumber.js';

import type { PrimeApy, Token } from 'types';
import {
  appendPrimeSimulationDistributions,
  convertAprBipsToApy,
  extractSettledPromiseValue,
  findTokenByAddress,
} from 'utilities';

import { getLegacyPoolMarkets, getTokenBalances } from 'clients/api';
import { formatToPool } from './formatToPool';
import type { GetLegacyPoolInput, GetLegacyPoolOutput } from './types';

export type { GetLegacyPoolInput, GetLegacyPoolOutput } from './types';

const getLegacyPool = async ({
  chainId,
  blocksPerDay,
  name,
  provider,
  description,
  xvs,
  vai,
  tokens,
  accountAddress,
  vTreasuryContractAddress,
  legacyPoolComptrollerContract,
  venusLensContract,
  vaiControllerContract,
  resilientOracleContract,
  primeContract,
}: GetLegacyPoolInput): Promise<GetLegacyPoolOutput> => {
  const [
    marketsResult,
    mainMarkets,
    xvsPriceMantissaResult,
    primeVTokenAddressesResult,
    primeMinimumXvsToStakeResult,
    userPrimeTokenResult,
    assetsInResult,
    _accrueVaiInterestResult,
    vaiRepayAmountResult,
  ] = await Promise.allSettled([
    // Fetch all markets
    legacyPoolComptrollerContract.getAllMarkets(),
    // Fetch main markets to get the supplier and borrower counts
    // TODO: fetch borrower and supplier counts from subgraph once available
    getLegacyPoolMarkets({ xvs }),
    // Fetch XVS price
    resilientOracleContract.getPrice(xvs.address),
    // Prime related calls
    primeContract?.getAllMarkets(),
    primeContract?.MINIMUM_STAKED_XVS(),
    // Account related calls
    accountAddress ? primeContract?.tokens(accountAddress) : undefined,
    accountAddress ? legacyPoolComptrollerContract.getAssetsIn(accountAddress) : undefined,
    // Call (statically) accrueVAIInterest to calculate past accrued interests before fetching all
    // interests. Since multicall will batch these requests, the call to accrueVAIInterest and
    // getVAIRepayAmount will happen in the same request (thus making the accrual possible)
    accountAddress ? vaiControllerContract.callStatic.accrueVAIInterest() : undefined,
    accountAddress ? vaiControllerContract.getVAIRepayAmount(accountAddress) : undefined,
  ]);

  if (marketsResult.status === 'rejected') {
    throw new Error(marketsResult.reason);
  }

  if (xvsPriceMantissaResult.status === 'rejected') {
    throw new Error(xvsPriceMantissaResult.reason);
  }

  const vTokenAddresses = marketsResult.value;
  const primeVTokenAddresses = extractSettledPromiseValue(primeVTokenAddressesResult) || [];
  const primeMinimumXvsToStakeMantissa = extractSettledPromiseValue(primeMinimumXvsToStakeResult);
  const isUserPrime = extractSettledPromiseValue(userPrimeTokenResult)?.exists || false;

  // Fetch vToken meta data
  const vTokenMetaDataPromise = venusLensContract.callStatic.vTokenMetadataAll(vTokenAddresses);

  // Fetch underlying token prices
  const underlyingTokenPricePromises = Promise.allSettled(
    vTokenAddresses.map(vTokenAddress => resilientOracleContract.getUnderlyingPrice(vTokenAddress)),
  );

  // Fetch vToken borrow and supply caps
  const borrowCapsPromises = Promise.allSettled(
    vTokenAddresses.map(vTokenAddress => legacyPoolComptrollerContract.borrowCaps(vTokenAddress)),
  );
  const supplyCapsPromises = Promise.allSettled(
    vTokenAddresses.map(vTokenAddress => legacyPoolComptrollerContract.supplyCaps(vTokenAddress)),
  );

  // Fetch vToken borrow and supply speeds
  const xvsBorrowSpeedPromises = Promise.allSettled(
    vTokenAddresses.map(vTokenAddress =>
      legacyPoolComptrollerContract.venusBorrowSpeeds(vTokenAddress),
    ),
  );
  const xvsSupplySpeedPromises = Promise.allSettled(
    vTokenAddresses.map(vTokenAddress =>
      legacyPoolComptrollerContract.venusSupplySpeeds(vTokenAddress),
    ),
  );

  // Fetch Prime distributions
  const primeAprPromises =
    primeContract && isUserPrime
      ? Promise.allSettled(
          accountAddress
            ? primeVTokenAddresses.map(primeVTokenAddress =>
                primeContract.calculateAPR(primeVTokenAddress, accountAddress),
              )
            : [],
        )
      : undefined;

  const underlyingTokenPriceResults = await underlyingTokenPricePromises;
  const borrowCapsResults = await borrowCapsPromises;
  const supplyCapsResults = await supplyCapsPromises;
  const xvsBorrowSpeedResults = await xvsBorrowSpeedPromises;
  const xvsSupplySpeedResults = await xvsSupplySpeedPromises;
  const vTokenMetaDataResults = await vTokenMetaDataPromise;
  const primeAprResults = (await primeAprPromises) || [];

  const primeApyMap = new Map<string, PrimeApy>();
  primeAprResults.forEach((primeAprResult, index) => {
    if (primeAprResult.status !== 'fulfilled') {
      return;
    }

    const primeApr = primeAprResult.value;

    const apys: PrimeApy = {
      borrowApy: convertAprBipsToApy({ aprBips: primeApr?.borrowAPR.toString() || '0' }),
      supplyApy: convertAprBipsToApy({ aprBips: primeApr?.supplyAPR.toString() || '0' }),
    };

    primeApyMap.set(primeVTokenAddresses[index], apys);
  });

  const vaiRepayAmountMantissa = extractSettledPromiseValue(vaiRepayAmountResult);

  const underlyingTokens = vTokenMetaDataResults.reduce<Token[]>((acc, vTokenMetaData) => {
    const underlyingToken = findTokenByAddress({
      address: vTokenMetaData.underlyingAssetAddress,
      tokens,
    });

    if (!underlyingToken) {
      return acc;
    }

    return [...acc, underlyingToken];
  }, []);

  // Fetch vToken meta data and user balance
  const [vTreasuryTokenBalancesResult, userVTokenBalancesResults] = await Promise.allSettled([
    // Fetch treasury balances
    getTokenBalances({
      accountAddress: vTreasuryContractAddress,
      tokens: underlyingTokens,
      provider,
    }),
    // Fetch user vToken balances
    accountAddress
      ? venusLensContract.callStatic.vTokenBalancesAll(vTokenAddresses, accountAddress)
      : undefined,
  ]);

  const pool = formatToPool({
    chainId,
    blocksPerDay,
    name,
    xvs,
    vai,
    tokens,
    description,
    comptrollerContractAddress: legacyPoolComptrollerContract.address,
    vTokenMetaDataResults,
    underlyingTokenPriceResults,
    borrowCapsResults,
    supplyCapsResults,
    xvsBorrowSpeedResults,
    xvsSupplySpeedResults,
    xvsPriceMantissa: new BigNumber(xvsPriceMantissaResult.value.toString()),
    userCollateralizedVTokenAddresses: extractSettledPromiseValue(assetsInResult),
    vTreasuryTokenBalances: extractSettledPromiseValue(vTreasuryTokenBalancesResult),
    userVTokenBalances: extractSettledPromiseValue(userVTokenBalancesResults),
    userVaiBorrowBalanceMantissa: vaiRepayAmountMantissa
      ? new BigNumber(vaiRepayAmountMantissa.toString())
      : undefined,
    primeApyMap,
    mainMarkets: extractSettledPromiseValue(mainMarkets)?.markets,
  });

  // Fetch Prime simulations and add them to distributions
  if (primeContract && primeMinimumXvsToStakeMantissa) {
    await appendPrimeSimulationDistributions({
      assets: pool.assets,
      primeContract,
      primeVTokenAddresses,
      accountAddress,
      primeMinimumXvsToStakeMantissa: new BigNumber(primeMinimumXvsToStakeMantissa.toString()),
      xvs,
    });
  }

  return {
    pool,
  };
};

export default getLegacyPool;
