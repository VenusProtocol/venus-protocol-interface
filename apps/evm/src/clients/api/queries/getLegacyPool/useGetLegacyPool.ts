import { type QueryObserverOptions, useQuery } from '@tanstack/react-query';

import getLegacyPool, {
  type GetLegacyPoolInput,
  type GetLegacyPoolOutput,
} from 'clients/api/queries/getLegacyPool';
import FunctionKey from 'constants/functionKey';
import { useGetChainMetadata } from 'hooks/useGetChainMetadata';
import { useIsFeatureEnabled } from 'hooks/useIsFeatureEnabled';
import {
  useGetLegacyPoolComptrollerContract,
  useGetPrimeContract,
  useGetResilientOracleContract,
  useGetVaiControllerContract,
  useGetVenusLensContract,
} from 'libs/contracts';
import { useGetToken, useGetTokens } from 'libs/tokens';
import { useTranslation } from 'libs/translations';
import { useChainId } from 'libs/wallet';
import type { ChainId } from 'types';
import { callOrThrow, generatePseudoRandomRefetchInterval } from 'utilities';

type TrimmedInput = Omit<
  GetLegacyPoolInput,
  | 'chainId'
  | 'blocksPerDay'
  | 'provider'
  | 'name'
  | 'description'
  | 'venusLensContract'
  | 'legacyPoolComptrollerContract'
  | 'resilientOracleContract'
  | 'vaiControllerContract'
  | 'vTreasuryContractAddress'
  | 'vai'
  | 'xvs'
  | 'tokens'
>;

export type UseGetLegacyPoolQueryKey = [
  FunctionKey.GET_LEGACY_POOL,
  TrimmedInput & { chainId: ChainId },
  additionalQueryKey?: 'PrimeCalculator',
];

type Options = QueryObserverOptions<
  GetLegacyPoolOutput,
  Error,
  GetLegacyPoolOutput,
  GetLegacyPoolOutput,
  UseGetLegacyPoolQueryKey
>;

const refetchInterval = generatePseudoRandomRefetchInterval();

const useGetLegacyPool = (input?: TrimmedInput, options?: Partial<Options>) => {
  const { chainId } = useChainId();
  const { blocksPerDay } = useGetChainMetadata();

  const { t } = useTranslation();

  const xvs = useGetToken({ symbol: 'XVS' });
  const vai = useGetToken({ symbol: 'VAI' });
  const tokens = useGetTokens();
  const isPrimeEnabled = useIsFeatureEnabled({
    name: 'prime',
  });

  const legacyPoolComptrollerContract = useGetLegacyPoolComptrollerContract();
  const venusLensContract = useGetVenusLensContract();
  const resilientOracleContract = useGetResilientOracleContract();
  const vaiControllerContract = useGetVaiControllerContract();
  const primeContract = useGetPrimeContract();

  const isQueryEnabled =
    !!legacyPoolComptrollerContract &&
    !!venusLensContract &&
    !!vai &&
    !!vaiControllerContract &&
    (options?.enabled === undefined || options?.enabled);

  return useQuery({
    queryKey: [FunctionKey.GET_LEGACY_POOL, { ...input, chainId }],
    queryFn: () =>
      callOrThrow(
        {
          xvs,
          legacyPoolComptrollerContract,
          venusLensContract,
          resilientOracleContract,
          vai,
          vaiControllerContract,
          blocksPerDay,
        },
        params =>
          getLegacyPool({
            chainId,
            name: t('legacyPool.name'),
            description: t('legacyPool.description'),
            tokens,
            primeContract: isPrimeEnabled ? primeContract : undefined,
            ...input,
            ...params,
          }),
      ),
    refetchInterval,
    ...options,
    enabled: isQueryEnabled,
  });
};

export default useGetLegacyPool;
