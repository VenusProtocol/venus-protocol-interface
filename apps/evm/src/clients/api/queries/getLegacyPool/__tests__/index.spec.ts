import fakeAccountAddress from '__mocks__/models/address';
import tokens, { vai, xvs } from '__mocks__/models/tokens';

import { ChainId } from 'types';
import getLegacyPool from '..';
import {
  fakeLegacyPoolComptrollerContract,
  fakeResilientOracleContract,
  fakeVaiControllerContract,
  fakeVenusLensContract,
} from '../__testUtils__/fakeData';

describe('getLegacyPool', () => {
  it('returns core pool in the correct format', async () => {
    const response = await getLegacyPool({
      chainId: ChainId.BSC_TESTNET,
      blocksPerDay: 28800,
      name: 'Fake pool name',
      description: 'Fake pool description',
      xvs,
      vai,
      tokens,
      accountAddress: fakeAccountAddress,
      legacyPoolComptrollerContract: fakeLegacyPoolComptrollerContract,
      venusLensContract: fakeVenusLensContract,
      vaiControllerContract: fakeVaiControllerContract,
      resilientOracleContract: fakeResilientOracleContract,
    });

    // Check VAI interests were accrued before being fetched
    expect(fakeVaiControllerContract.callStatic.accrueVAIInterest).toHaveBeenCalledTimes(1);
    expect(response).toMatchSnapshot();
  });
});
