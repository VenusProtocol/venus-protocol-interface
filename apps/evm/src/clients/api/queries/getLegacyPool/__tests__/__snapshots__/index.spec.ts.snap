// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getLegacyPool > returns core pool in the correct format 1`] = `
{
  "pool": {
    "assets": [
      {
        "borrowApyPercentage": "2.0200781054088957",
        "borrowBalanceCents": "24663011775918.58670630114424219627258",
        "borrowBalanceTokens": "246628588661.936163058800477859",
        "borrowCapTokens": "1e-18",
        "borrowDistributions": [
          {
            "apyPercentage": "0",
            "dailyDistributedTokens": "0",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "borrowerCount": 438,
        "cashTokens": "1.00000000080503544824460983238917793784e+21",
        "collateralFactor": 0,
        "disabledTokenActions": [
          "supply",
          "borrow",
        ],
        "exchangeRateVTokens": "49.19405324154271215086",
        "isCollateralOfUser": true,
        "liquidityCents": "1.000006200805040439464388948970138750743214608e+23",
        "reserveFactor": 1,
        "reserveTokens": "0",
        "supplierCount": 933,
        "supplyApyPercentage": "0",
        "supplyBalanceCents": "1.000006201051670556767362816225462034303610544746e+23",
        "supplyBalanceTokens": "1.00000000105166403645033682413715238395883e+21",
        "supplyCapTokens": "0",
        "supplyDistributions": [
          {
            "apyPercentage": "0",
            "dailyDistributedTokens": "0",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "tokenPriceCents": "100.00062",
        "userBorrowBalanceCents": "1.02596026091418e-9",
        "userBorrowBalanceTokens": "1.0259539e-11",
        "userPercentOfLimit": 0,
        "userSupplyBalanceCents": "0",
        "userSupplyBalanceTokens": "0",
        "userWalletBalanceCents": "0.00213276018859808651478",
        "userWalletBalanceTokens": "0.000021327469655669",
        "vToken": {
          "address": "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4",
          "decimals": 8,
          "symbol": "vBUSD",
          "underlyingToken": {
            "address": "0x8301F2213c0eeD49a7E28Ae4c3e91722919B8B47",
            "asset": "/src/libs/tokens/img/busd.svg",
            "decimals": 18,
            "symbol": "BUSD",
          },
        },
      },
      {
        "borrowApyPercentage": "12.763322565556457",
        "borrowBalanceCents": "3721959398.320962344455482972",
        "borrowBalanceTokens": "15136.746248595758022748",
        "borrowCapTokens": "0",
        "borrowDistributions": [
          {
            "apyPercentage": "1.5797413696681861",
            "dailyDistributedTokens": "250.0000128",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "borrowerCount": 35,
        "cashTokens": "0",
        "collateralFactor": 0.8,
        "disabledTokenActions": [],
        "exchangeRateVTokens": "37.63966306652044917955",
        "isCollateralOfUser": false,
        "liquidityCents": "0",
        "reserveFactor": 0.1,
        "reserveTokens": "21.269535066688975633",
        "supplierCount": 25,
        "supplyApyPercentage": "11.434048167920574",
        "supplyBalanceCents": "3716729453.6129492577232450137",
        "supplyBalanceTokens": "15115.4767135290690422233",
        "supplyCapTokens": "1.15792089237316195423570985008687907853269984665640564039457584007913129639935e+59",
        "supplyDistributions": [
          {
            "apyPercentage": "1.5819817239252298",
            "dailyDistributedTokens": "250.0000128",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "tokenPriceCents": "245889",
        "userBorrowBalanceCents": "7.156865078267903019",
        "userBorrowBalanceTokens": "0.000029106080704171",
        "userPercentOfLimit": 0,
        "userSupplyBalanceCents": "0",
        "userSupplyBalanceTokens": "0",
        "userWalletBalanceCents": "2458865411100005.406193157416225272",
        "userWalletBalanceTokens": "9999900000.000021986315603448",
        "vToken": {
          "address": "0x162D005F0Fff510E54958Cfc5CF32A3180A84aab",
          "decimals": 8,
          "symbol": "vETH",
          "underlyingToken": {
            "address": "0x98f7A83361F7Ac8765CcEBAB1425da6b341958a7",
            "asset": "/src/libs/tokens/img/eth.svg",
            "decimals": 18,
            "isNative": true,
            "symbol": "ETH",
          },
        },
      },
      {
        "borrowApyPercentage": "180.75799822550164",
        "borrowBalanceCents": "54206766464.991794941213834066224",
        "borrowBalanceTokens": "1032405.72373577234796371",
        "borrowCapTokens": "0",
        "borrowDistributions": [
          {
            "apyPercentage": "0.6478221277312901",
            "dailyDistributedTokens": "1499.9999904",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "borrowerCount": 263,
        "cashTokens": "5.556233762724731858",
        "collateralFactor": 0.8,
        "disabledTokenActions": [],
        "exchangeRateVTokens": "5.63429988023545433776",
        "isCollateralOfUser": true,
        "liquidityCents": "291731.6894670817923653489952",
        "reserveFactor": 0.25,
        "reserveTokens": "187487.447405339178551222",
        "supplierCount": 923,
        "supplyApyPercentage": "157.57792083092227",
        "supplyBalanceCents": "44362974574.359412320127072323564288",
        "supplyBalanceTokens": "844923.83256419589412162352",
        "supplyCapTokens": "1.15792089237316195423570985008687907853269984665640564039457584007913129639935e+59",
        "supplyDistributions": [
          {
            "apyPercentage": "0.7921351276470645",
            "dailyDistributedTokens": "1499.9999904",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "tokenPriceCents": "52505.2944",
        "userBorrowBalanceCents": "7073785.2980067928964376533856",
        "userBorrowBalanceTokens": "134.725181123959051574",
        "userPercentOfLimit": 0.12,
        "userSupplyBalanceCents": "6813717433.5010155085170052265088",
        "userSupplyBalanceTokens": "129772.007020705620660552",
        "userWalletBalanceCents": "540373271888369.14416",
        "userWalletBalanceTokens": "10291786343.9",
        "vToken": {
          "address": "0x2E7222e51c0f6e98610A1543Aa3836E092CDe62c",
          "decimals": 8,
          "symbol": "vBNB",
          "underlyingToken": {
            "address": "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
            "asset": "/src/libs/tokens/img/bnb.svg",
            "decimals": 18,
            "isNative": true,
            "symbol": "BNB",
          },
        },
      },
      {
        "borrowApyPercentage": "16.850090258739982",
        "borrowBalanceCents": "266527277.02974427316207663041589",
        "borrowBalanceTokens": "416884.364702270734855123",
        "borrowCapTokens": "270000",
        "borrowDistributions": [
          {
            "apyPercentage": "0",
            "dailyDistributedTokens": "0",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "borrowerCount": 68,
        "cashTokens": "2259657.074761148481496107",
        "collateralFactor": 0.6,
        "disabledTokenActions": [],
        "exchangeRateVTokens": "47.36910235297224095005",
        "isCollateralOfUser": true,
        "liquidityCents": "1444669788.91666196711723462774301",
        "reserveFactor": 0.25,
        "reserveTokens": "0",
        "supplierCount": 274,
        "supplyApyPercentage": "1.8360754649567212",
        "supplyBalanceCents": "1711197065.9464062401284554003137589",
        "supplyBalanceTokens": "2676541.43946341921611527123",
        "supplyCapTokens": "3750000",
        "supplyDistributions": [
          {
            "apyPercentage": "0",
            "dailyDistributedTokens": "0",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "tokenPriceCents": "639.33143",
        "userBorrowBalanceCents": "0",
        "userBorrowBalanceTokens": "0",
        "userPercentOfLimit": 0,
        "userSupplyBalanceCents": "0.70725092468251742803237",
        "userSupplyBalanceTokens": "0.001106235188034659",
        "userWalletBalanceCents": "6.3933143000000639331429999863775169751254466605905e+26",
        "userWalletBalanceTokens": "1.000000000000009999999999997869261171021335e+24",
        "vToken": {
          "address": "0x6d6F697e34145Bb95c54E77482d97cc261Dc237E",
          "decimals": 8,
          "symbol": "vXVS",
          "underlyingToken": {
            "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
            "asset": "/src/libs/tokens/img/xvs.svg",
            "decimals": 18,
            "symbol": "XVS",
          },
        },
      },
      {
        "borrowApyPercentage": "2.0200781032592374",
        "borrowBalanceCents": "743724749.881571973423",
        "borrowBalanceTokens": "7437523.951581",
        "borrowCapTokens": "0",
        "borrowDistributions": [
          {
            "apyPercentage": "16.981964056809385",
            "dailyDistributedTokens": "499.9999968",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "borrowerCount": 249,
        "cashTokens": "1.0000736633870942814396398228131475074669e+34",
        "collateralFactor": 0.8,
        "disabledTokenActions": [],
        "exchangeRateVTokens": "49.8066352191466231613",
        "isCollateralOfUser": false,
        "liquidityCents": "1.000036490649026183145198711400933542774047455327e+36",
        "reserveFactor": 0.1,
        "reserveTokens": "0",
        "supplierCount": 670,
        "supplyApyPercentage": "0",
        "supplyBalanceCents": "1.00003649064902618314530970887489046343322854178897705712904885e+36",
        "supplyBalanceTokens": "1.000073663387094281439750824413033895903139261475326095e+34",
        "supplyCapTokens": "1.15792089237316195423570985008687907853269984665640564039457584007913129639935e+71",
        "supplyDistributions": [
          {
            "apyPercentage": "0",
            "dailyDistributedTokens": "499.9999968",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "tokenPriceCents": "99.996283",
        "userBorrowBalanceCents": "0",
        "userBorrowBalanceTokens": "0",
        "userPercentOfLimit": 0,
        "userSupplyBalanceCents": "0",
        "userSupplyBalanceTokens": "0",
        "userWalletBalanceCents": "9999928289169302585.414550788886",
        "userWalletBalanceTokens": "100003000003203144.915042",
        "vToken": {
          "address": "0xb7526572FFE56AB9D7489838Bf2E18e3323b441A",
          "decimals": 8,
          "symbol": "vUSDT",
          "underlyingToken": {
            "address": "0xA11c8D9DC9b66E209Ef60F0C8D969D3CD988782c",
            "asset": "/src/libs/tokens/img/usdt.svg",
            "decimals": 6,
            "symbol": "USDT",
          },
        },
      },
      {
        "borrowApyPercentage": "2.0200782609031798",
        "borrowBalanceCents": "1562397934365776.075433634816",
        "borrowBalanceTokens": "15624370577897.031296",
        "borrowCapTokens": "0",
        "borrowDistributions": [
          {
            "apyPercentage": "0.000007467880158706",
            "dailyDistributedTokens": "499.9999968",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "borrowerCount": 229,
        "cashTokens": "1.010176805025969148511141552e+21",
        "collateralFactor": 0.81,
        "disabledTokenActions": [],
        "exchangeRateVTokens": "47.12928141003911013496",
        "isCollateralOfUser": true,
        "liquidityCents": "1.01015151019877129824366283301553792e+23",
        "reserveFactor": 0.1,
        "reserveTokens": "0",
        "supplierCount": 534,
        "supplyApyPercentage": "2.7328805884e-8",
        "supplyBalanceCents": "1.0101515258227468345608256850251328456045167117888e+23",
        "supplyBalanceTokens": "1.01017682065033591897223674983334867265528e+21",
        "supplyCapTokens": "1.15792089237316195423570985008687907853269984665640564039457584007913129639935e+71",
        "supplyDistributions": [
          {
            "apyPercentage": "0",
            "dailyDistributedTokens": "499.9999968",
            "token": {
              "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
              "asset": "/src/libs/tokens/img/xvs.svg",
              "decimals": 18,
              "symbol": "XVS",
            },
            "type": "rewardDistributor",
          },
        ],
        "tokenPriceCents": "99.997496",
        "userBorrowBalanceCents": "0",
        "userBorrowBalanceTokens": "0",
        "userPercentOfLimit": 0,
        "userSupplyBalanceCents": "0",
        "userSupplyBalanceTokens": "0",
        "userWalletBalanceCents": "0",
        "userWalletBalanceTokens": "0",
        "vToken": {
          "address": "0xD5C4C2e2facBEB59D0216D0595d63FcDc6F9A1a7",
          "decimals": 8,
          "symbol": "vUSDC",
          "underlyingToken": {
            "address": "0x16227D60f7a0e586C66B005219dfc887D13C9531",
            "asset": "/src/libs/tokens/img/usdc.svg",
            "decimals": 6,
            "symbol": "USDC",
          },
        },
      },
    ],
    "comptrollerAddress": "0xa258a693A403b7e98fd05EE9e1558C760308cFC7",
    "description": "Fake pool description",
    "isIsolated": false,
    "name": "Fake pool name",
    "userBorrowBalanceCents": "7073792.45487187219030093329978",
    "userBorrowLimitCents": "5450973947.225162961623114638026462",
    "userSupplyBalanceCents": "6813717434.20826643319952265454117",
  },
}
`;
